l1-contracts/contracts/bridgehub/Bridgehub.sol

```diff
@@ -0,0 +1,249 @@
+        _;
+    }
+
+    modifier onlystateTransitionManagerChain(uint256 _chainId) {
+        require(
+            msg.sender == IStateTransitionManager(stateTransitionManager[_chainId]).stateTransition(_chainId),
+            "Bridgehub: not state transition chain"
+        );
+        _;
+    }
+
+    modifier onlyBaseTokenBridge(uint256 _chainId) {
+        require(msg.sender == baseTokenBridge[_chainId], "Bridgehub: not base token bridge");
+        _;
+    }
+
+    function initialize() external reentrancyGuardInitializer returns (bytes32) {
+    }
+
+    ///// Getters
+
+    function getZkSyncStateTransition(uint256 _chainId) public view returns (address) {
+        return IStateTransitionManager(stateTransitionManager[_chainId]).stateTransition(_chainId);
+    }
+
+    //// Registry
+
+    /// @notice State Transition can be any contract with the appropriate interface/functionality
+    function newstateTransitionManager(address _stateTransitionManager) external onlyOwner {
+        require(!stateTransitionManagerIsRegistered[_stateTransitionManager], "Bridgehub: state transition already registered");
+        stateTransitionManagerIsRegistered[_stateTransitionManager] = true;
+    }
+
+    /// @notice State Transition can be any contract with the appropriate interface/functionality
+    /// @notice this stops new Chains from using the STF, old chains are not affected
+    function removestateTransitionManager(address _stateTransitionManager) external onlyOwner {
+        require(stateTransitionManagerIsRegistered[_stateTransitionManager], "Bridgehub: state transition already registered");
+        stateTransitionManagerIsRegistered[_stateTransitionManager] = false;
+    }
+
+    /// @notice token can be any contract with the appropriate interface/functionality
+    function newToken(address _token) external onlyOwner {
+        require(!tokenIsRegistered[_token], "Bridgehub: token already registered");
+        tokenIsRegistered[_token] = true;
+    }
+
+    /// @notice Bridge can be any contract with the appropriate interface/functionality
+    function newTokenBridge(address _tokenBridge) external onlyOwner {
+        require(!tokenBridgeIsRegistered[_tokenBridge], "Bridgehub: token bridge already registered");
+        tokenBridgeIsRegistered[_tokenBridge] = true;
+    }
+
+    function setWethBridge(address _wethBridge) external onlyOwner {
+        wethBridge = IL1Bridge(_wethBridge);
+    }
+
+    /// @notice for Eth the baseToken address is 0.
+    function newChain(
+        uint256 _chainId,
+        address _stateTransitionManager,
+        address _baseToken,
+        address _baseTokenBridge,
+        uint256 _salt,
+        address _l2Governor,
+        bytes calldata _initData
+    ) external onlyOwner returns (uint256 chainId) {
+        // KL TODO: clear up this formula for chainId generation
+        if (_chainId == 0) {
+            chainId = uint48(
```

Comment:
I would rather remove chainID in flight generation for now. It won't be used in the initial release, so better to not add it as dead code.

Reply:
There is a bug already with this functionality. In the line above `_chainId` is zero, while real chain id should be checked.

```solidity
require(stateTransitionManager[_chainId] == address(0), "Bridgehub: chainId already registered");
```

Reply:
True, this should be chainId instead of _chainId

Reply:
But I think we should have permissinoless deployment of hyperchains on testnet. So for that reason I would keep this

Reply:
Ok, removed that functionality now